quic/s2n-quic-core/src/recovery/bbr/pacing.rs

```diff
@@ -0,0 +1,297 @@
+    /// Sets the pacing rate used for determining the earliest departure time
+    #[inline]
+    pub(super) fn set_pacing_rate(&mut self, bw: Bandwidth, gain: Ratio<u64>, filled_pipe: bool) {
+        //= https://tools.ietf.org/id/draft-cardwell-iccrg-bbr-congestion-control-02#2.5
+        //# The static discount factor of 1% used to scale BBR.bw to produce BBR.pacing_rate.
+        const PACING_MARGIN_PERCENT: u64 = 1;
+        const PACING_RATIO: Ratio<u64> = Ratio::new_raw(100 - PACING_MARGIN_PERCENT, 100);
+
+        //= https://tools.ietf.org/id/draft-cardwell-iccrg-bbr-congestion-control-02#4.6.2
+        //# BBRSetPacingRateWithGain(pacing_gain):
+        //#   rate = pacing_gain * bw * (100 - BBRPacingMarginPercent) / 100
+        //#   if (BBR.filled_pipe || rate > BBR.pacing_rate)
+        //#     BBR.pacing_rate = rate
+        let rate = bw * gain * PACING_RATIO;
+
+        if filled_pipe || rate > self.pacing_rate {
+            self.pacing_rate = rate;
+        }
+    }
+
+    /// Sets the maximum size of data aggregate scheduled and transmitted together
+    #[inline]
+    pub(super) fn set_send_quantum(&mut self, max_datagram_size: u16) {
+        //= https://tools.ietf.org/id/draft-cardwell-iccrg-bbr-congestion-control-02#4.6.3
+        //# if (BBR.pacing_rate < 1.2 Mbps)
+        //#   floor = 1 * SMSS
+        //# else
+        //#   floor = 2 * SMSS
+        //# BBR.send_quantum = min(BBR.pacing_rate * 1ms, 64KBytes)
+        //# BBR.send_quantum = max(BBR.send_quantum, floor)
+
+        // 1.2 Mbps
+        const SEND_QUANTUM_THRESHOLD: Bandwidth =
+            Bandwidth::new(1_200_000 / 8, Duration::from_secs(1));
+
+        let floor = if self.pacing_rate < SEND_QUANTUM_THRESHOLD {
+            max_datagram_size
+        } else {
+            max_datagram_size * 2
+        } as usize;
+
+        let send_quantum = (self.pacing_rate * Duration::from_millis(1)) as usize;
+        self.send_quantum = send_quantum
+            .max(floor)
+            .min(Self::max_send_quantum(max_datagram_size));
+    }
+
+    /// Returns the earliest time that a packet may be transmitted.
+    ///
+    /// If the time is in the past or is `None`, the packet should be transmitted immediately.
+    pub(super) fn earliest_departure_time(&self) -> Option<Timestamp> {
+        self.next_packet_departure_time
+    }
+
+    /// Returns the maximum size of data aggregate scheduled and transmitted together
+    pub(super) fn send_quantum(&self) -> usize {
+        self.send_quantum
+    }
+
+    /// Returns the maximum value for send_quantum
+    #[inline]
+    fn max_send_quantum(max_datagram_size: u16) -> usize {
+        //= https://tools.ietf.org/id/draft-cardwell-iccrg-bbr-congestion-control-02#4.6.3
+        //= type=exception
+        //= reason=QUIC recommends limiting bursts to the initial congestion window
+        //# BBR.send_quantum = min(BBR.pacing_rate * 1ms, 64KBytes)
+        MAX_BURST_PACKETS as usize * max_datagram_size as usize
```

Comment:
Do we plan on ever raising this value? In theory, the BBR model could track quantum delivery rate and shrink it if the path doesn't handle the bursty traffic (I have no idea if it does or not).

Also do we need to consider raising the GSO limits?

https://github.com/aws/s2n-quic/blob/38409d931165cf36656c06b5f7e5c2a0fcb46997/quic/s2n-quic-platform/src/features/gso.rs#L73

Reply:
BBR sets the send quantum based on the pacing rate, which is in term based on the maximum bandwidth sample. So if bandwidth is reduced due to the path not handling bursty traffic, that could have an effect on the send quantum. 

We could consider raising the limits to see if the reduction in CPU is worth any negative effects in more burstiness, though the BBR spec recommends going as small as possible:

> A BBR implementation MAY use alternate approaches to select a BBR.send_quantum, as appropriate for the CPU overheads anticipated for senders and receivers, and buffering considerations anticipated in the network path. However, for the sake of the network and other users, a BBR implementation SHOULD attempt to use the smallest feasible quanta.